### PR TITLE
Docker for mac take 2

### PR DESCRIPTION
**Pull request by csantanapr**
This change is a continuation of PR #2440 and #2524
To give it a try
```
sudo ifconfig lo0 alias 172.17.0.1/24
```
Using `wskdev`
```
wskdev fresh -t local
```
Using `ansible-playbook`
```
cd ansible
ansible-playbook -i environments/local setup.yml
ansible-playbook -i environments/local couchdb.yml
ansible-playbook -i environments/local initdb.yml
ansible-playbook -i environments/local wipe.yml
ansible-playbook -i environments/local apigateway.yml
ansible-playbook -i environments/local openwhisk.yml
ansible-playbook -i environments/local postdeploy.yml
```

No need to `sudo` on the mac
Addresses also `become: true`

### CODE REVIEW

File: ansible/logs.yml [hunk 1]
```diff
@@ -24,14 +24,13 @@
         - subjects
       when: "'db' not in exclude_logs_from"
 
-- hosts: all:!ansible
-  serial: 1
+- hosts: shared,core,invokers
   tasks:
   - name: get all docker containers
-    local_action: shell docker --host tcp://{{ ansible_host }}:{{ docker.port }} ps -a --format="{% raw %}{{.Names}}{% endraw %}"
```
Comment (rabbah) [inline ansible/logs.yml]: isn't this semantic changing?

Comment (csantanapr) [reply ansible/logs.yml]: Good catch, I will leave it working as is now for linux and docker-machine and use `--host tcp://{{ ansible_host }}:{{ docker.port }}` and for Docker for Mac it will not present.

**Comment by csantanapr** [comment]
@rabbah addressed your requested changes

File: ansible/README.md [hunk 2]
```diff
@@ -30,18 +30,8 @@ You need to depend on the workarounds until Docker provides official methods.
...
-
-There are many workarounds for this.
-One way is to use `socat` command to setup proxy for the UNIX socket.
-
-```
-docker run -d -v /var/run/docker.sock:/var/run/docker.sock -p 4243:2375 bobrik/socat TCP4-LISTEN:2375,fork,reuseaddr UNIX-CONNECT:/var/run/docker.sock
-```
-
 If you want to deploy OpenWhisk in distributed environment, you are required to enable Docker remote API in all Mac hosts.
+Only required when collecting logs using the Ansible playbook [logs.yml](logs.yml)
```
Comment (rabbah) [inline ansible/README.md]: so we don't need the remote api except for logs? I'd reword the two sentences then.

File: ansible/README.md [hunk 3]
```diff
@@ -74,8 +64,9 @@ systemProp.http.proxyPort=3128
 **Caveat:** All Ansible commands are meant to be executed from the `ansible` directory.
 This is important because that's where `ansible.cfg` is located which contains generic settings that are needed for the remaining steps.
 
-In all instructions, replace `<environment>` with your target environment. e.g. `mac`if you want to deploy using a local mac setup.
+In all instructions, replace `<environment>` with your target environment. e.g. `docker-machine`if you want to deploy using a local docker-machine setup.
```
Comment (rabbah) [inline ansible/README.md]: should we make docker-machine secondary?

File: ansible/logs.yml [hunk 4]
```diff
@@ -27,11 +27,19 @@
 - hosts: all:!ansible
   serial: 1
   tasks:
+  - name: set host flag
+    set_fact:
+      docker_host_flag: "--host tcp://{{ ansible_host }}:{{ docker.port }}"
+    when: (whisk_version_name == "mac" or ansible_os_family == "Debian")
```
Comment (rabbah) [inline ansible/logs.yml]: will this work in our pgs?

File: tests/src/test/scala/actionContainers/ActionContainer.scala [hunk 5]
```diff
@@ -134,30 +155,34 @@ object ActionContainer {
...
 
     try {
+      //TODO: look into this, need it if not it throws java.lang.IllegalStateException
+      Thread.sleep(500)
       // ...and finally run the code with it.
       code(mock)
       // I'm told this is good for the logs.
       Thread.sleep(100)
-      val (_, out, err) = awaitDocker(s"logs $name", 10 seconds)
+      val (_, out, err) = awaitDocker(s" logs $name", 10 seconds)
```
Comment (rabbah) [inline tests/src/test/scala/actionContainers/ActionContainer.scala]: why the spaces before the sub commands?

**Comment by csantanapr** [comment]
PG3 1080 🔵

File: ansible/logs.yml (hunk 4 above)
Comment (csantanapr) [reply ansible/logs.yml]: yes works on pgs

File: ansible/README.md (hunk 3 above)
Comment (csantanapr) [reply ansible/README.md]: I don't know what you mean, proposed text to add instead

File: ansible/README.md (hunk 2 above)
Comment (csantanapr) [reply ansible/README.md]: provide alternative text if required change

Comment (rabbah) [reply ansible/README.md]: suggestion:
```
-If you want to deploy OpenWhisk in distributed environment, you are required to enable Docker remote API in all Mac hosts.
-Only required when collecting logs using the Ansible playbook [logs.yml](logs.yml)
+The remote Docker API is required for collecting logs using the Ansible playbook [logs.yml](logs.yml).
```

File: .gitignore [hunk 6]
```diff
@@ -50,6 +50,7 @@ node_modules
 .idea
 *.class
 *.iml
+out/
```
Comment (rabbah) [inline .gitignore]: for?

File: ansible/README.md (hunk 3 above)
Comment (rabbah) [reply ansible/README.md]: suggested patch:

```
-In all instructions, replace `<environment>` with your target environment. e.g. `docker-machine`if you want to deploy using a local docker-machine setup.
-By default, if you omit the `-i` parameter, the `local` environment will be used.
-If using Docker for Mac use the `local` environment.
+In all instructions, replace `<environment>` with your target environment. The default environment is `local` which works for Ubuntu and
+Docker for Mac. To use the default environment, you may omit the `-i` parameter entirely. For older Mac installation using Docker Machine,
+use `-i environments/docker-machine`.
```

File: ansible/README.md [hunk 7]
```diff
@@ -125,7 +116,7 @@ ansible-playbook -i environments/<environment> setup.yml
 ```
 
 #### Install Prerequisites
-This step needs to be done only once per target environment (in a local setup the development environment and the target environment are the same). It will install necessary prerequisites on all target hosts in the environment.
+This step needs to be done only once per target environment. It will install necessary prerequisites on all target hosts in the environment. No required for local environments since all prerequisites are already installed.
```
Comment (rabbah) [inline ansible/README.md]: `No required for local environments since all prerequisites are already installed.` ->
`This step is not required for local environments since all prerequisites are already installed, and therefore may be skipped.`

File: ansible/logs.yml [hunk 8]
```diff
@@ -27,11 +27,19 @@
...
+  - name: do not set host flag when Docker mac
+    set_fact:
+      docker_host_flag: ""
+    when: (whisk_version_name != "mac" and ansible_os_family != "Debian")
   - name: get all docker containers
-    local_action: shell docker --host tcp://{{ ansible_host }}:{{ docker.port }} ps -a --format="{% raw %}{{.Names}}{% endraw %}"
+    local_action: shell docker  {{ docker_host_flag }} ps -a --format="{% raw %}{{.Names}}{% endraw %}"
```
Comment (rabbah) [inline ansible/logs.yml]: extra space after `docker` here and line 42.

File: ansible/roles/couchdb/tasks/deploy.yml [hunk 9]
```diff
@@ -41,12 +41,13 @@
       COUCHDB_PASSWORD: "{{ db_password }}"
       NODENAME: "{{ nodeName }}"
 
-- name: wait until the CouchDB in this host is up and running
-  wait_for:
-    delay: 2
-    host: "{{ ansible_host }}"
-    port: "{{ db_port }}"
-    timeout: 60
+- name: wait until CouchDB in this host is up and running
```
Comment (rabbah) [inline ansible/roles/couchdb/tasks/deploy.yml]: wasn't this part of a different pr?

File: tools/build/redo [hunk 10]
```diff
@@ -205,14 +205,11 @@ Components = [
     makeComponent('fresh',
                   'setup, build, and deploy a fresh whisk system using couchdb',
                   yaml = False,
-                  steps = 'setup, prereq, couchdb, initdb, wipedb, deploy, catalog'),
+                  steps = 'setup, couchdb, initdb, wipedb, deploy, catalog'),
 
     makeComponent('setup',
                   'system setup'),
 
-    makeComponent('prereq',
```
Comment (rabbah) [inline tools/build/redo]: keep this but remove it from fresh?

File: ansible/logs.yml (hunk 4 above)
Comment (rabbah) [reply ansible/logs.yml]: confirmed that it did work on the distributed environment... but i'm a little puzzled why --- can you help me understand why this works on gnu/linux?

File: ansible/roles/couchdb/tasks/deploy.yml (hunk 9 above)
Comment (rabbah) [reply ansible/roles/couchdb/tasks/deploy.yml]: yes it was https://github.com/apache/incubator-openwhisk/commit/602ce9438cdf5cbfa5e3365223c1c48348ff1092.
just need a rebase.

File: ansible/roles/invoker/tasks/deploy.yml [hunk 11]
```diff
@@ -40,23 +56,43 @@
...
     linuxOptions: "-v /usr/lib/x86_64-linux-gnu/libapparmor.so.1:/usr/lib/x86_64-linux-gnu/libapparmor.so.1"
-  when: whisk_version_name != "mac"
+  when: ansible_distribution == "Ubuntu"
 
 - name: get running invoker information
-  uri:
-    url: "http://{{ ansible_host }}:{{ docker.port }}/containers/json?filters={{ '{\"name\":[ \"invoker\" ],\"ancestor\":[ \"invoker\" ]}' | urlencode }}"
-    return_content: yes
-  register: invokerInfo
+  uri: url="http://{{ inventory_hostname }}:{{ docker.port }}/containers/json?filters={{ '{"name":[ "invoker" ]}' | urlencode }}" return_content=yes
```
Comment (rabbah) [inline ansible/roles/invoker/tasks/deploy.yml]: losing the `ancestor` filter is that deliberate?

File: core/actionProxy/Dockerfile [hunk 12]
```diff
@@ -19,5 +19,5 @@ ADD actionproxy.py /actionProxy/
 RUN mkdir -p /action
 ADD stub.sh /action/exec
 RUN chmod +x /action/exec
-
+EXPOSE 8080
```
Comment (rabbah) [inline core/actionProxy/Dockerfile]: since we didn't address the proxy issue yet for the local container tests, is exposing the port in the runtime containers necessary?

**Comment by csantanapr** [comment]
PG1 2061 🔵  with latest commit

File: ansible/logs.yml (hunk 4 above)
Comment (csantanapr) [reply ansible/logs.yml]: because this`ansible_os_family == "Debian"` will be `true` in those environments

Comment (csantanapr) [reply ansible/logs.yml]: Once we remove `docker-machine` this will look cleaner

File: .gitignore (hunk 6 above)
Comment (csantanapr) [reply .gitignore]: intelliJ

File: tests/src/test/scala/actionContainers/ActionContainer.scala [hunk 13]
```diff
@@ -90,7 +90,9 @@ object ActionContainer {
   }
 
   private lazy val dockerCmd: String = {
-    val hostStr = if (WhiskProperties.onMacOSX()) {
+    val version = WhiskProperties.getProperty("whisk.version.name")
+    //check if we are running on docker-machine env
```
Comment (rabbah) [inline tests/src/test/scala/actionContainers/ActionContainer.scala]: format comment - i know you like to make me happy :)

File: ansible/logs.yml (hunk 4 above)
Comment (rabbah) [reply ansible/logs.yml]: wondering if we can condition this on `!local`.

File: core/actionProxy/Dockerfile (hunk 12 above)
Comment (csantanapr) [reply core/actionProxy/Dockerfile]: not necessary

**Comment by csantanapr** [comment]
Playground3 1093 🔵

File: ansible/logs.yml (hunk 4 above)
Comment (csantanapr) [reply ansible/logs.yml]: ah missed some other places

**Comment by csantanapr** [comment]
Playground3 1095 🚀

File: tools/build/redo [hunk 14]
```diff
@@ -215,7 +215,7 @@ Components = [
 
     makeComponent('setup',
                   'system setup'),
-
+
```
Comment (rabbah) [inline tools/build/redo]: surprised this passed scan code.

Comment (rabbah) [reply tools/build/redo]: i wonder if we’re exempting redo. will have to look.

**Comment by rabbah** [comment]
I failed deploying on docker-machine after rebasing this branch.

```
TASK [invoker : determine docker root dir on docker-machine] ********************************************************************************************************************************************************************************
Friday 15 September 2017  12:22:34 -0400 (0:00:00.034)       0:00:44.740 ****** 
fatal: [invoker0]: FAILED! => {"changed": false, "content": "", "failed": true, "msg": "Status code was not [200]: Request failed: <urlopen error timed out>", "redirected": false, "status": -1, "url": "http://invoker0:4243/info"}
fatal: [invoker1]: FAILED! => {"changed": false, "content": "", "failed": true, "msg": "Status code was not [200]: Request failed: <urlopen error timed out>", "redirected": false, "status": -1, "url": "http://invoker1:4243/info"}

PLAY RECAP **********************************************************************************************************************************************************************************************************************************
192.168.99.100             : ok=10   changed=9    unreachable=0    failed=0   
controller0                : ok=4    changed=2    unreachable=0    failed=0   
controller1                : ok=4    changed=2    unreachable=0    failed=0   
invoker0                   : ok=1    changed=0    unreachable=0    failed=1   
invoker1                   : ok=1    changed=0    unreachable=0    failed=1  
```

a docker for mac deployment was fine.

**Comment by csantanapr** [comment]
I think I figure out what was wrong

**Comment by csantanapr** [comment]
Playground4 797 🔵 🕺

**Comment by csantanapr** [comment]
Playground1 2065 🔵 🎉
